### PR TITLE
change generators to IDL-based pipeline

### PR DESCRIPTION
**Pull request by dirk-thomas**
Connected to ros2/rosidl#334.

### CODE REVIEW

File: rosidl_generator_py/rosidl_generator_py/generate_py_impl.py [hunk 1]
```diff
@@ -19,11 +19,30 @@
...
+        'int8': 'int8_t',
+        'uint16': 'uint16_t',
+        'int16': 'int16_t',
+        'uint32': 'uint32_t',
+        'int32': 'int32_t',
+        'uint64': 'uint64_t',
+        'int64': 'int64_t',
+        'string': 'rosidl_generator_c__String',
+    }
+    return msg_type_to_c[type_]
```
Comment (hidmic) [inline rosidl_generator_py/rosidl_generator_py/generate_py_impl.py]: @dirk-thomas I've seen this mapping replicated in multiple places, here and in https://github.com/ros2/rosidl/pull/334, but I'm sure I'm missing a part of the history.

Comment (dirk-thomas) [reply rosidl_generator_py/rosidl_generator_py/generate_py_impl.py]: The mapping shouldn't be identical. They are in general specific to the generator.

File: rosidl_generator_py/resource/_idl_pkg_typesupport_entry_point.c.em [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+@#######################################################################
+@# Handle actions
+@#######################################################################
+@{
+from rosidl_parser.definition import Action
+}@
+@[for action in content.get_elements_of_type(Action)]@
+
+@{
+TEMPLATE(
```
[the remark this answers is not in the document]
Comment (dirk-thomas) [reply rosidl_generator_py/resource/_idl_pkg_typesupport_entry_point.c.em]: Moved in 6fe24416101ef5f0a5710f1c52a394d359fd8c9a.

File: rosidl_generator_py/resource/_idl_pkg_typesupport_entry_point.c.em [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+  pymodule = PyModule_Create(&@(package_name)__module);
+  if (!pymodule) {
+    return NULL;
+  }
+  int8_t err;
+@[for register_function in register_functions]@
+
+  err = @(register_function)(pymodule);
+  if (err) {
+    Py_XDECREF(pymodule);
```
[the remark this answers is not in the document]
Comment (dirk-thomas) [reply rosidl_generator_py/resource/_idl_pkg_typesupport_entry_point.c.em]: What variable are you referring to / do you want to be decref-ed?

File: rosidl_generator_py/resource/_msg.py.em [hunk 4]
```diff
@@ -40,148 +41,193 @@ class Metaclass(type):
...
+importable_typesupports = set()
+for member in message.structure.members:
+    type_ = member.type
+    if isinstance(type_, NestedType):
+        type_ = type_.basetype
+    if isinstance(type_, NamespacedType):
+        if (
+            type_.name.endswith(ACTION_GOAL_SUFFIX) or
+            type_.name.endswith(ACTION_RESULT_SUFFIX) or
+            type_.name.endswith(ACTION_FEEDBACK_SUFFIX)
```
[the remark this answers is not in the document]
Comment (dirk-thomas) [reply rosidl_generator_py/resource/_msg.py.em]: The wrapping data structures will have field with the type of the goal / result / feedback. But the import for those is using only the action name (without the `_Goal` / `_Result` / `_Feedback` suffix).

File: rosidl_generator_py/cmake/rosidl_generator_py_generate_interfaces.cmake [hunk 5]
```diff
@@ -34,83 +34,44 @@ endif()
...
-    )
-  else()
-    message(FATAL_ERROR "Interface file with unknown parent folder: ${_idl_file}")
-  endif()
+  get_filename_component(_idl_name "${_abs_idl_file}" NAME_WE)
+  string_camel_case_to_lower_case_underscore("${_idl_name}" _module_name)
+  list(APPEND _generated_py_files
+    "${_output_path}/${_parent_folder}/_${_module_name}.py")
+  list(APPEND _generated_c_files
+    "${_output_path}/${_parent_folder}/_${_module_name}_s.c")
```
[the remark this answers is not in the document]
Comment (dirk-thomas) [reply rosidl_generator_py/cmake/rosidl_generator_py_generate_interfaces.cmake]: I don't know what that suffix was chosen when this was implemented originally in 1d7b4eb15a0fae579ec4e02f1345156886235200.

File: rosidl_generator_py/cmake/rosidl_generator_py_generate_interfaces.cmake [hunk 6]
```diff
@@ -34,83 +34,44 @@ endif()
...
-endif()
+set(_generated_py_dirs "")
+foreach(_generated_py_file ${_generated_py_files})
+  get_filename_component(_parent_folder "${_generated_py_file}" DIRECTORY)
+  set(_init_module "${_parent_folder}/__init__.py")
+  list(FIND _generated_py_files "${_init_module}" _index)
+  if(_index EQUAL -1)
+    list(APPEND _generated_py_files "${_init_module}")
+
+    string(LENGTH "${_output_path}" _length)
```
[the remark this answers is not in the document]
Comment (dirk-thomas) [reply rosidl_generator_py/cmake/rosidl_generator_py_generate_interfaces.cmake]: Added in ed35aaba1ec02f0e9314dc205a5d55a1c0d44496.

File: rosidl_generator_py/resource/_msg.py.em (hunk 4 above)
[the remark this answers is not in the document]
Comment (jacobperron) [reply rosidl_generator_py/resource/_msg.py.em]: Is there a risk that a user defines a regular message that collides with the action suffix (e.g. `MyMessage_Feedback`)?

File: rosidl_generator_py/resource/_idl_pkg_typesupport_entry_point.c.em (hunk 3 above)
[the remark this answers is not in the document]
Comment (jacobperron) [reply rosidl_generator_py/resource/_idl_pkg_typesupport_entry_point.c.em]: nevermind, I was thinking that the register functions were PyObjects.

File: rosidl_generator_py/resource/_msg.py.em (hunk 4 above)
[the remark this answers is not in the document]
Comment (dirk-thomas) [reply rosidl_generator_py/resource/_msg.py.em]: No, underscores are not allowed in a message type name.

**Pull request merged**